SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,273 @@
+	}
+	else
+	{
+		CurrentSpanId = Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	}
+
+	Trace_Event TraceEvent{ CurrentSpanId, /* unix_timestamp_millis: ignored */ 0, EventMessage.GetMessage(), EventMessage.GetType(),
+							nullptr };
+	if (!Trace_EventTracer_ShouldSampleEvent(EventTracer, &TraceEvent))
+	{
+		return {};
+	}
+
+	Trace_EventData* EventData = Trace_EventData_Create();
+
+	auto AddTraceEventStringField = [EventData](const FString& KeyString, const FString& ValueString) {
+		auto KeySrc = StringCast<ANSICHAR>(*KeyString);
+		const ANSICHAR* Key = KeySrc.Get();
+		auto ValueSrc = StringCast<ANSICHAR>(*ValueString);
+		const ANSICHAR* Value = ValueSrc.Get();
+		Trace_EventData_AddStringFields(EventData, 1, &Key, &Value);
+	};
+
+#if ENGINE_MINOR_VERSION >= 25
+	using UnrealProperty = FProperty;
+#else
+	using UnrealProperty = UProperty;
+#endif
+
+	for (TFieldIterator<UnrealProperty> It(Struct); It; ++It)
+	{
+		UnrealProperty* Property = *It;
+
+		FString VariableName = Property->GetName();
+		const void* Value = Property->ContainerPtrToValuePtr<uint8>(&EventMessage);
+
+		check(Property->ArrayDim == 1); // Arrays not handled yet
+
+#if ENGINE_MINOR_VERSION >= 25
+		if (FStrProperty* StringProperty = CastField<FStrProperty>(Property))
+#else
+		if (UStrProperty* StringProperty = Cast<UStrProperty>(Property))
+#endif
+		{
+			AddTraceEventStringField(VariableName, StringProperty->GetPropertyValue(Value));
+		}
+#if ENGINE_MINOR_VERSION >= 25
+		else if (FObjectProperty* ObjectProperty = CastField<FObjectProperty>(Property))
+#else
+		else if (UObjectProperty* ObjectProperty = Cast<UObjectProperty>(Property))
+#endif
+		{
+			UObject* Object = ObjectProperty->GetPropertyValue(Value);
+			if (Object)
+			{
+				AddTraceEventStringField(VariableName, Object->GetName());
+				if (AActor* Actor = Cast<AActor>(Object))
+				{
+					FString KeyString = VariableName + TEXT("Position");
+					FString ValueString = Actor->GetTransform().GetTranslation().ToString();
+					AddTraceEventStringField(KeyString, ValueString);
+				}
+			}
+			else
+			{
+				AddTraceEventStringField(VariableName, "Null");
+			}
+		}
+		else // Default
+		{
+			FString StringValue;
+			Property->ExportTextItem(StringValue, Value, NULL, NULL, PPF_None);
+			AddTraceEventStringField(VariableName, StringValue);
+		}
+	}
+
+	// TODO(EventTracer): implement and call AddTargetObjectInfoToEventData
+	// TODO(EventTracer): implement and call AddComponentIdInfoToEventData
+
+	TraceEvent.data = EventData;
+	Trace_EventTracer_AddEvent(EventTracer, &TraceEvent);
+	Trace_EventData_Destroy(EventData);
+
+	return CurrentSpanId;
+}
+
+bool SpatialEventTracer::IsEnabled() const
+{
+	return bEnabled; // Trace_EventTracer_IsEnabled(EventTracer);
+}
+
+void SpatialEventTracer::Enable(const FString& FileName)
+{
+	Trace_EventTracer_Parameters parameters = {};
+	EventTracerGuard.EventTracer = this;
```

Comment:
Add a scope lock for this as well (a callback from the previous tracer may get triggered just as the new one is being made, also we might want to pass the pointer as user data after all and compare in the callback that it's the same one).

Reply:
![image](https://user-images.githubusercontent.com/56311103/93220293-35d28200-f764-11ea-98c0-649732046dbd.png)

Like this?

Reply:
Almost, except I think it's clearer to do `reinterpret_cast<SpatialEventTracer*>(UserData)` because that's what happens behind the scenes since `SpatialEventTracer` is not a `UObject` (and if it was, it wouldn't be safe to call `Cast` in this scenario)